modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,385 @@
+use crate::error::AppError;
+use abstract_dex_adapter::msg::OfferAsset;
+use abstract_sdk::features::AbstractResponse;
+use cosmwasm_std::{DepsMut, Env, MessageInfo, Response, StdError, Uint128};
+
+use crate::contract::{AppResult, ChallengeApp};
+use abstract_sdk::prelude::*;
+use chrono::{Datelike, NaiveDateTime};
+
+use crate::msg::ChallengeExecuteMsg;
+use crate::state::{
+    ChallengeEntry, CheckIn, Friend, Penalty, Vote, CHALLENGE_FRIENDS, CHALLENGE_LIST,
+    DAILY_CHECK_INS, NEXT_ID, VOTES,
+};
+
+pub fn execute_handler(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    app: ChallengeApp,
+    msg: ChallengeExecuteMsg,
+) -> AppResult {
+    match msg {
+        ChallengeExecuteMsg::CreateChallenge { challenge } => {
+            create_challenge(deps, env, info, app, challenge)
+        }
+        ChallengeExecuteMsg::UpdateChallenge {
+            challenge_id,
+            challenge,
+        } => update_challenge(deps, env, info, app, challenge_id, challenge),
+        ChallengeExecuteMsg::CancelChallenge { challenge_id } => {
+            cancel_challenge(deps, info, app, challenge_id)
+        }
+        ChallengeExecuteMsg::AddFriendForChallenge {
+            challenge_id,
+            friend_name,
+            friend_address,
+        } => add_friend_for_challenge(
+            deps,
+            info,
+            &app,
+            challenge_id,
+            &friend_name,
+            &friend_address,
+        ),
+        ChallengeExecuteMsg::RemoveFriendForChallenge {
+            challenge_id,
+            friend_address,
+        } => remove_friend_from_challenge(deps, info, &app, challenge_id, friend_address),
+        ChallengeExecuteMsg::AddFriendsForChallenge {
+            challenge_id,
+            friends,
+        } => add_friends_for_challenge(deps, info, &app, challenge_id, friends),
+        ChallengeExecuteMsg::DailyCheckIn {
+            challenge_id,
+            metadata,
+        } => daily_check_in(deps, env, info, &app, challenge_id, metadata),
+        ChallengeExecuteMsg::CastVote { vote, challenge_id } => {
+            cast_vote(deps, env, info, &app, vote, &challenge_id)
+        }
+        ChallengeExecuteMsg::CountVotes { challenge_id } => {
+            count_votes(deps, info, env, &app, challenge_id)
+        }
+    }
+}
+
+/// Create new challenge
+fn create_challenge(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    app: ChallengeApp,
+    challenge: ChallengeEntry,
+) -> AppResult {
+    // Only the admin should be able to create a challenge.
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    // Generate the challenge id
+    let id = NEXT_ID.update(deps.storage, |id| AppResult::Ok(id + 1))?;
+    let challenge_id = format!("challenge_{id}");
```

Comment:
What's the reason for formatting this into a fixed string? I'd either just stick to a number (takes less storage) or let the user customize the ID as a string on creation (which could be nice!)

Reply:
I went with generated`u64` id, less storage is better and less error prone with potential malicious user input, possible dups and validation.